File: leetcode/medium/BestTimeToBuyAndSellStockII.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """
        Calculates the maximum profit that can be obtained by buying and selling a stock multiple times.

        The problem allows for multiple transactions (buy one share, sell it, buy again).
        The key insight is that the total profit can be calculated by accumulating the profit
        from every consecutive day where the price increases. This is a greedy approach.

        For example, if prices are [1, 5, 8], buying at 1 and selling at 8 (profit 7)
        is the same as buying at 1, selling at 5 (profit 4), and then immediately buying
        at 5 and selling at 8 (profit 3). Total profit = 4 + 3 = 7.

        This simplifies the problem to just summing up all positive price differences
        between consecutive days.

        Args:
            prices: A list of integers representing the stock price on consecutive days.

        Returns:
            The maximum possible profit.
        """
        # Handle the edge case where there are fewer than 2 days to make a transaction.
        if not prices or len(prices) < 2:
            return 0

        max_profit = 0
        # Iterate through the prices starting from the second day.
        for i in range(1, len(prices)):
            # If the current day's price is higher than the previous day's price,
            # it represents a profitable transaction (buy yesterday, sell today).
            if prices[i] > prices[i - 1]:
                max_profit += prices[i] - prices[i - 1]

        return max_profit

File: leetcode/medium/BestTimeToBuyAndSellStockII.java
/**
 * Solution for the Best Time to Buy and Sell Stock II problem.
 */
class Solution {
    /**
     * Calculates the maximum profit that can be obtained by buying and selling a stock multiple times.
     *
     * The problem allows for completing as many transactions as you like (i.e., buy one and sell one share
     * of the stock multiple times). However, you may not engage in multiple transactions simultaneously
     * (i.e., you must sell the stock before you buy again).
     *
     * The strategy used here is a greedy approach. We iterate through the prices and accumulate
     * profit whenever the price on the current day is higher than the price on the previous day.
     * This is equivalent to buying on the low days and selling on the high days. For example,
     * if prices are [1, 2, 3], adding (2-1) + (3-2) = 2 is the same as buying at 1 and selling at 3.
     *
     * @param prices An array of integers where prices[i] is the price of the stock on day i.
     * @return The maximum profit that can be achieved.
     */
    public int maxProfit(int[] prices) {
        // If the prices array is null or has fewer than 2 elements, no transaction is possible.
        if (prices == null || prices.length < 2) {
            return 0;
        }

        int totalProfit = 0;

        // Iterate through the prices starting from the second day.
        for (int i = 1; i < prices.length; i++) {
            // If the price on the current day is greater than the previous day,
            // we can make a profit.
            if (prices[i] > prices[i - 1]) {
                // Add the difference to the total profit. This simulates buying on day i-1
                // and selling on day i.
                totalProfit += prices[i] - prices[i - 1];
            }
        }

        // Return the accumulated profit.
        return totalProfit;
    }
}
